akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -162,23 +170,31 @@ private[http] object OutgoingConnectionBlueprint {
       }
 
       def onPull(): Unit = {
-        if (!entitySubstreamStarted) pull(in)
+        if (!entitySubstreamStarted && !hasBeenPulled(in)) pull(in)
```

Comment:
I’d fold this change into the waitForMessageEnd handler instead (by making it extend OutHandler as well and ignoring onPull there)